### PR TITLE
Make classic scoring a constant multiple of standardised scoring

### PR DESCRIPTION
**Pull request by smoogipoo**
In order to fix score reordering on the leaderboards, we need standardised scoring and classic scoring to directly correlate to each other. As far as I'm aware, this can only be done if classic scoring is a constant multiple of standardised scoring, which is what I've changed it to, based on max achievable combo.

[Freedom Dive](https://osu.ppy.sh/beatmapsets/39804#osu/129891) comparison: 

| player | stable | master | This PR |
| --- | --- | --- | --- |
| [chocomint](https://osu.ppy.sh/scores/osu/2177560145) | 132M | 77M | 159M |
| [Reedkatt](https://osu.ppy.sh/scores/osu/3188565717) | 67M | 49M | 117M |
| [smoogipoo](https://osu.ppy.sh/scores/osu/1001256491) | 1.6M | 2M | 18M |

Other maps are similar results - it's accurate at or very close to full-combo, and less accurate at lesser plays. But, if this is an issue, it is an issue with standardised scoring also.

### CODE REVIEW

**Comment by bdach** [comment]
I'm not sure if this will help at all but I'm pretty sure that the requirement for classic scoring to be a constant multiple of standardised is too strict.

Assuming that classic scoring is to be calculated off of standardised via a single-variable function `f(x)`, then I *think* that for order to be preserved, it is sufficient for `f` to be an increasing function over its entire domain. (Non-decreasing should also work, I guess.) Constant multiple meets this criteria, obviously.

So in human language, I'm reasonably certain that something like using an exponential should work just as well to preserve order.

**Comment by peppy** [comment]
@bdach i think that was the gist of the conversation between smoogi and i regarding this (there is likely a better way) but probably easiest if you are able make a quick proposal code suggestion as to how you'd see this working. the issue with the previous algorithm was the combo was affecting the accuracy portion of the score, creating a cross-dependency that could change order.

**Comment by smoogipoo** [comment]
Well, it's exponential in terms of number of hitobjects. I consider that as a "constant multiple" since it doesn't change between plays on the same beatmap, even though it's not a numeric constant.

Am I misunderstanding? I couldn't find a way to relate e.g. accuracy and combo as it was previously being done.

**Comment by bdach** [comment]
@smoogipoo Exactly, it's a constant multiple as is for one given beatmap. What I was trying to get at is that you could try to approximate *the rate* of ascent as score goes up by using the exponential. Because standardised goes up relatively flat (it's all percentages and ratios, iirc), the classic scoring will look more "flat" too with this change with the way they're correlated. Using an exponential in there (applied to the standardised score) could shake that up and make classic "feel" more like it used to be.

Obviously it is impossible to exactly reproduce the previous way, as you've both rightly stated, since the variables are interdependent.

@peppy Sure, I can have a go later. Just wanted to get that piece of info out now that there may be a method that preserves the spirit of classic more.

**Comment by smoogipoo** [comment]
Right, so it would be an exponential on the entire first part of the equation (`standardisedScore / max_score`). Give it a shot, this can change at any point since it's only visual.

**Comment by bdach** [comment]
I've tried out something like the following:

```diff
diff --git a/osu.Game/Rulesets/Scoring/ScoreProcessor.cs b/osu.Game/Rulesets/Scoring/ScoreProcessor.cs
index 2a7691269d..e09225f967 100644
--- a/osu.Game/Rulesets/Scoring/ScoreProcessor.cs
+++ b/osu.Game/Rulesets/Scoring/ScoreProcessor.cs
@@ -227,7 +227,8 @@ public double GetScore(ScoringMode mode, int maxCombo, double accuracyRatio, dou
                 case ScoringMode.Classic:
                     // This gives a similar feeling to osu!stable scoring (ScoreV1) while keeping classic scoring as only a constant multiple of standardised scoring.
                     // The invariant is important to ensure that scores don't get re-ordered on leaderboards between the two scoring modes.
-                    return GetScore(ScoringMode.Standardised, maxCombo, accuracyRatio, comboRatio, statistics) / max_score * Math.Pow(maxCombo, 2) * 25;
+                    var scaledStandardised = GetScore(ScoringMode.Standardised, maxCombo, accuracyRatio, comboRatio, statistics) / max_score;
+                    return Math.Pow(scaledStandardised * maxCombo, 2) * 18;
             }
         }
```

<details>
<summary>Nerd details why (skip to see the results)</summary>
<p>
It's not an exponential, but a quadratic. The quadratic is still an increasing function (at least for positive arguments), and the reason why I chose that rather than exp is because standardised scoring is a linear mix of the combo and accuracy portions, which are themselves *mostly* just the combo and accuracy values but rescaled linearly. So the old classic algorithm could *sorta kinda* be interpreted as quadratically correlated to standardised if you squint enough.
</p>
<p>
(Also the quadratic has a nice property that if you square something in the [0,1] interval, you are _guaranteed_ to get a number from the [0,1] interval back as the result. It's a bit less apparent in the version above since the scaled thing is pulled into the `Math.Pow()` call, but as you may recall, *a*<sup>2</sup>*b*<sup>2</sup> = (*ab*)<sup>2</sup>.)
</p>
The 25 -> 18 change is just to scale a bit closer to stable. More on that below.
</details>

Results from my (rather rudimentary) testing:

### Freedom Dive (from the OP)

| | stable | master | this PR as-is | my proposal |
| :- | -: | -: | -: | -: |
| [chocomint](https://osu.ppy.sh/scores/osu/2177560145) | 132,408,001 | 77,248,748 | 159,701,664 | 129,132,216 |
| [Reedkatt](https://osu.ppy.sh/scores/osu/3188565717) | 67,290,028 | 49,257,045 | 117,595,086 | 70,015,480 |

I couldn't run this on @smoogipoo's score as I simply don't have it, so I approximated the "bad player" case myself via a single play on a map I can actually play without utterly dying, which is:

### [Boxplot - Tramontane \[Reform's Insane\]](https://osu.ppy.sh/beatmapsets/1016398#osu/2359368)

| | stable | master | this PR as-is | my proposal |
| :- | -: | -: | -: | -: |
| [Mathi](https://osu.ppy.sh/scores/osu/3436941098) | 54,779,319 | 40,418,540 | 83,322,124 | 71,181,170 |
| [xarrian](https://osu.ppy.sh/scores/osu/3162770979) | 46,088,988 | 34,102,394 | 70,170,173 | 50,483,511 |
| me | - | 5,259,089 | 13,754,195 | 5,076,254 |

**Comment by smoogipoo** [comment]
That looks much better.

**Comment by peppy** [comment]
@smoogipoo want to apply @bdach's change? and probably check on the test failures.

File: osu.Game.Tests/Rulesets/Scoring/ScoreProcessorTest.cs [hunk 1]
```diff
@@ -85,19 +85,18 @@ public void TestSingleOsuHit(ScoringMode scoringMode, HitResult hitResult, int e
...
-        // TODO: The following two cases don't match expectations currently (a single hit is registered in acc portion when it shouldn't be). See https://github.com/ppy/osu/issues/12604.
-        [TestCase(ScoringMode.Classic, HitResult.SmallBonus, HitResult.SmallBonus, 330)] // (1 * 1 * 300) * (1 + 0 / 25) + 3 * 10 (bonus points)
-        [TestCase(ScoringMode.Classic, HitResult.LargeBonus, HitResult.LargeBonus, 450)] // (1 * 1 * 300) * (1 + 0 / 25) + 3 * 50 (bonus points)
+        [TestCase(ScoringMode.Classic, HitResult.Miss, HitResult.Great, 0)]
+        [TestCase(ScoringMode.Classic, HitResult.Meh, HitResult.Great, 68)]
+        [TestCase(ScoringMode.Classic, HitResult.Ok, HitResult.Great, 81)]
+        [TestCase(ScoringMode.Classic, HitResult.Good, HitResult.Perfect, 109)]
+        [TestCase(ScoringMode.Classic, HitResult.Great, HitResult.Great, 149)]
+        [TestCase(ScoringMode.Classic, HitResult.Perfect, HitResult.Perfect, 149)]
+        [TestCase(ScoringMode.Classic, HitResult.SmallTickMiss, HitResult.SmallTickHit, 9)]
```
Comment (smoogipoo) [inline osu.Game.Tests/Rulesets/Scoring/ScoreProcessorTest.cs:94]: Note: This changed from 0 to 9. But it matches the non-zero standardised form above.

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs [hunk 2]
```diff
@@ -222,12 +222,13 @@ public double GetScore(ScoringMode mode, int maxCombo, double accuracyRatio, dou
...
-
                     return (max_score * (accuracyScore + comboScore) + getBonusScore(statistics)) * scoreMultiplier;
 
                 case ScoringMode.Classic:
-                    // should emulate osu-stable's scoring as closely as we can (https://osu.ppy.sh/help/wiki/Score/ScoreV1)
-                    return getBonusScore(statistics) + (accuracyRatio * Math.Max(1, maxCombo) * 300) * (1 + Math.Max(0, (comboRatio * maxCombo) - 1) * scoreMultiplier / 25);
+                    // This gives a similar feeling to osu!stable scoring (ScoreV1) while keeping classic scoring as only a constant multiple of standardised scoring.
+                    // The invariant is important to ensure that scores don't get re-ordered on leaderboards between the two scoring modes.
+                    double scaledStandardised = GetScore(ScoringMode.Standardised, maxCombo, accuracyRatio, comboRatio, statistics) / max_score;
+                    return Math.Pow(scaledStandardised * (maxCombo + 1), 2) * 18;
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs:231]: I added 1 to this to fix the only-bonus-score case. It'll change scores by a bit but I don't think it's relevant enough to reconsider the `18` multiplier.

Comment (bdach) [reply osu.Game/Rulesets/Scoring/ScoreProcessor.cs:231]: The difference in the inner term will not exceed `2 * maxCombo + 1`, so in the large scale of things this should not have a significant impact on the final numbers.

**Comment by bdach** [comment]
This was discussed in discord already but just for the sake of information parity, do note that by squaring the standardised score which already has mod multipliers applied, classic scoring basically has the mod multipliers squared. This cannot be rectified however without sacrificing stability of score order between classic and standardised.

**Review by peppy — APPROVED** [review]

**Pull request merged**